### PR TITLE
feat: OMV integration & health monitoring widget

### PR DESCRIPTION
**Pull request by hillaliy**
<br/>
<div align="center">
  
  <h3>Homarr</h3>
</div>

**Thank you for your contribution. Please ensure that your pull request meets the following pull request:**

- [x] Builds without warnings or errors (``pnpm buid``, autofix with ``pnpm format:fix``)
- [x] Pull request targets ``dev`` branch
- [x] Commits follow the [conventional commits guideline](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] No shorthand variable names are used (eg. ``x``, ``y``, ``i`` or any abbrevation)

### CODE REVIEW

File: packages/integrations/src/openmediavault/openmediavault-integration.ts
Comment (manuel-rw) [inline packages/integrations/src/openmediavault/openmediavault-integration.ts:15]: You never re-use this variable, inline it with line 16

File: packages/integrations/src/openmediavault/openmediavault-integration.ts
Comment (manuel-rw) [inline packages/integrations/src/openmediavault/openmediavault-integration.ts:40]: ``typeof result === "object"`` should make ``result !== null`` obsolete, no?

Comment (manuel-rw) [inline packages/integrations/src/openmediavault/openmediavault-integration.ts:40]: Invert to simplify condition?

File: packages/integrations/src/openmediavault/openmediavault-integration.ts
Comment (manuel-rw) [inline packages/integrations/src/openmediavault/openmediavault-integration.ts:61]: Extract this code to a private function?

File: packages/integrations/src/openmediavault/openmediavault-integration.ts
Comment (manuel-rw) [inline packages/integrations/src/openmediavault/openmediavault-integration.ts:91]: Rename ``fsys`` -> do not use abbreviations in names!

File: packages/integrations/src/openmediavault/openmediavault-integration.ts
Comment (manuel-rw) [inline packages/integrations/src/openmediavault/openmediavault-integration.ts:98]: Rename ``smt``. Do not use abbreviations

**Review by manuel-rw — CHANGES_REQUESTED** [review]

File: packages/integrations/src/openmediavault/openmediavault-integration.ts
Comment (Meierschlumpf) [reply packages/integrations/src/openmediavault/openmediavault-integration.ts:40]: No sadly not, `null` is an `object` in terms of typescript :joy:

File: packages/integrations/src/openmediavault/openmediavault-integration.ts
Comment (hillaliy) [reply packages/integrations/src/openmediavault/openmediavault-integration.ts:61]: when i make it private there is deepSource error

File: packages/integrations/src/openmediavault/openmediavault-integration.ts
Comment (hillaliy) [reply packages/integrations/src/openmediavault/openmediavault-integration.ts:40]: So the condition is OK?

File: packages/cron-jobs/src/jobs/integrations/health-monitoring.ts
Comment (manuel-rw) [inline packages/cron-jobs/src/jobs/integrations/health-monitoring.ts:39]: Are you aware that you never publish this information to Redis?

**Review by manuel-rw — CHANGES_REQUESTED** [review]

**Comment by manuel-rw** [comment]
@homarr-labs/maintainers please have a look at this

**Comment by SeDemal** [comment]
Don't forget to add a classnaming system for easy custom CSS.

Regarding the scaling, I don't have omv myself to test it out, could you make a recording of the widget and how it behaves when you resize it?

**Comment by hillaliy** [comment]
https://github.com/user-attachments/assets/aa9ecbe1-67dd-4de2-86c8-e2bee322aa79

**Comment by hillaliy** [comment]
> Don't forget to add a classnaming system for easy custom CSS.
> 
> Regarding the scaling, I don't have omv myself to test it out, could you make a recording of the widget and how it behaves when you resize it?

Already added

**Comment by SeDemal** [comment]
> > Don't forget to add a classnaming system for easy custom CSS.
> > Regarding the scaling, I don't have omv myself to test it out, could you make a recording of the widget and how it behaves when you resize it?
> 
> Already added

Either you haven't, or you forgot to push it. None of the components in the component.tsx file have className.

Also, the video shows that not all element are scaling, did you apply scaling to the text and icons?

**Comment by hillaliy** [comment]
https://github.com/user-attachments/assets/60ac9983-6b1d-46be-beed-cc5ef8e8f438

Thanks guys for all comments. I have learned a lot from this Integration / Widget.

File: packages/widgets/src/health-monitoring/component.tsx
Comment (manuel-rw) [inline packages/widgets/src/health-monitoring/component.tsx:85]: This icon looks too big to me. Do you agree?
Maybe make it slightly smaller?

File: packages/widgets/src/health-monitoring/component.tsx
Comment (manuel-rw) [inline packages/widgets/src/health-monitoring/component.tsx:122]: You should put the number and the ``:`` also in a translation.
In Arabic and other languages they will be reversed - hence they should be in the translations

**Review by manuel-rw — CHANGES_REQUESTED** [review]
Looks very good! Well done

**Comment by manuel-rw** [comment]
Also, please fix your merge conflicts and make sure that all actions are passing. Thank you!

File: packages/widgets/src/health-monitoring/component.tsx
Comment (hillaliy) [reply packages/widgets/src/health-monitoring/component.tsx:85]: I resize it to "8cqmin". looks better now

**Comment by manuel-rw** [comment]
As requested, I had a look at the build errors and fixed them for you in https://github.com/homarr-labs/homarr/pull/1142/commits/d1a40d686e28ce26436c19accc186e640fc2c47b.
The following fixes were required:
- You used a merge commit to merge the lastest changes from ``dev`` into your branch in https://github.com/homarr-labs/homarr/pull/1142/commits/846ba932b14694e11cc84fafe65ba3410c53b671. In this commit some functions and type definitions were renamed / refactored - this created merge conflicts
- You didn't resolve the merge conflicts correctly. You overwrote the changes from ``dev`` with yours - but in this case we wanted to combine both changes and apply them both.
- You should use the new functions to construct & transform data in the cron jobs:
image
- You overwrote some important changes in the integration definitions. I rolled them back for you:
image
Remember, **always read merge conflicts carefully** and evaluate what changes are more recent. Sometimes you'll need to manually combine or resolve conflicts because you can't use either one nor the combination of them directly.

I strongly suggest you to watch some guides on this topic. You did really well on this PR but I think you lack some Git knowledge.
My suggestion: https://www.youtube.com/watch?v=HosPml1qkrg

**Comment by manuel-rw** [comment]
There are some more errors that I overlooked. Will look into them as well...

File: packages/api/src/router/widgets/health-monitoring.ts
Comment (manuel-rw) [inline packages/api/src/router/widgets/health-monitoring.ts:21]: Do not call functions, that call integration APIs, in the TPC procedures.
We no longer do this because...
- this will not scale well with many users. Imagine that every client is invoking this procedure every 5 seconds. You have 100 users with 2 tabs of Homarr opened a the same time. You'll get approx. 1'000 API requests per minute which you will forward to the integration. Many integrations will break or refuse this many requests
- the ``await`` keyword will let the Promise resolve. This is bad because the procedure will have to wait for it to resolve. If the API request to the integration takes 8 seconds due to slow network, the procedure will be fetching for 8 seconds. This can lead to lags and suboptimal performance
- We want to pre-fetch data so data is already available before the user loads the page.
- We want to handle fetching asynchronously and handle errors in a more streamlined approach

As a summary; please move this call to the job and get the latest value from the in-memory Redis DB.

**Review by manuel-rw — CHANGES_REQUESTED** [review]

**Comment by SeDemal** [comment]
> /

That's a lot better, good improvement, congrats on getting there, and thanks for sticking with it.
I'll need to take a very close look at the code though.

Regarding the class-naming, Although it's tedious and ugly in the code, when I said to name every element, I meant it.
This is really important if we want to make CSS as easy for users as possible, especially those that know nothing about it.
We want to be able to give a name and directly affect it with CSS. Straight to the point.

For advanced information, we're now doing them in modals as hover cards don't scale nicely.
Those don't need class-naming though so you're fine there.

If it's as simple as a title though, just do a Tooltip.

in recap, if it's not inside a modal or a Tooltip, put a className on it.

File: packages/api/src/router/widgets/health-monitoring.ts
Comment (manuel-rw) [inline packages/api/src/router/widgets/health-monitoring.ts:17]: Variable doesn't add any value. Remove it and use ``healthInfo: data?.data`` where you return the object

File: packages/integrations/src/openmediavault/openmediavault-integration.ts
Comment (manuel-rw) [inline packages/integrations/src/openmediavault/openmediavault-integration.ts:7]: Private functions should be after public functions

File: packages/integrations/src/openmediavault/openmediavault-integration.ts
Comment (manuel-rw) [inline packages/integrations/src/openmediavault/openmediavault-integration.ts:27]: I dislike the name of this function. It doesn't really describe what happens inside it.
Suggestions: ``authenticateAndConstructSessionInHeaderAsync``, ``getAuthenticationHeaderAsync``, ...

Comment (manuel-rw) [inline packages/integrations/src/openmediavault/openmediavault-integration.ts:27]: Is it really required to call this function every request?
You will create double the noise for every API call. It isn't a bad thing but would be nice to avoid unnessesary network traffic if possible. Does the Cookie have an expiry?
If yes, maybe you can only perform this logic if the Cookie expired or will expire soon?
@Meierschlumpf I think your guidance would be nice here. Would a cache make sense here? And if yes, how do we deal with integration updates (eg. change URL)?

File: packages/old-import/src/widgets/options.ts
Comment (manuel-rw) [inline packages/old-import/src/widgets/options.ts:111]: We had more options in the old Homarr. Will you implement them in a follow up PR? Or are they no longer required?

File: packages/widgets/src/health-monitoring/component.tsx
Comment (manuel-rw) [inline packages/widgets/src/health-monitoring/component.tsx:64]: Extract

File: packages/widgets/src/health-monitoring/component.tsx
Comment (manuel-rw) [inline packages/widgets/src/health-monitoring/component.tsx:60]: ``uptime`` is in what unit? Suggestion: Add unit as a suffix to specify.
Example: ``ellapsedTimeSinceShutdownInMilliseconds``

**Review by manuel-rw — CHANGES_REQUESTED** [review]

File: packages/old-import/src/widgets/options.ts
Comment (hillaliy) [reply packages/old-import/src/widgets/options.ts:111]: The other options from old Homarr is for Proxmox integration. So it will take care when Proxmox integration will created

File: packages/integrations/src/openmediavault/openmediavault-integration.ts
Comment (hillaliy) [reply packages/integrations/src/openmediavault/openmediavault-integration.ts:27]: I found a nice solution.

**Review by manuel-rw — CHANGES_REQUESTED** [review]
@Meierschlumpf / @SeDemal can you take another look? LGTM

File: packages/translation/src/lang/en.ts
Comment (Meierschlumpf) [inline packages/translation/src/lang/en.ts:1097]: Maybe instead use the plurals feature from our translation library:
https://next-international.vercel.app/docs/app-plurals

File: packages/translation/src/lang/en.ts
Comment (Meierschlumpf) [inline packages/translation/src/lang/en.ts:1103]: Is this one used or can it be removed?

File: packages/widgets/src/health-monitoring/component.tsx
Comment (Meierschlumpf) [inline packages/widgets/src/health-monitoring/component.tsx:45]: This should be put underneath all hooks (So underneath the `useListState`. Otherwise when count is updated from 0 to 1 it will show a react error `More hooks rendered then in previous render`

File: packages/widgets/src/health-monitoring/component.tsx
Comment (Meierschlumpf) [inline packages/widgets/src/health-monitoring/component.tsx:143]: After the change in the translation file you could also make something similar to:

```ts
{[1,5,15].map(minutes => (
  <List.Item className={"load-average-${minutes}min"}>
    {t("widget.healthMonitoring.popover.minutes", {count: minutes})}{" "}{healthInfo.loadAverage[`${minutes}min`]} 
 </List.Item>
));
```

**Review by Meierschlumpf — CHANGES_REQUESTED** [review]

File: packages/integrations/src/openmediavault/openmediavault-integration.ts
Comment (hillaliy) [reply packages/integrations/src/openmediavault/openmediavault-integration.ts:7]: Public return deepSource error

Comment (hillaliy) [reply packages/integrations/src/openmediavault/openmediavault-integration.ts:7]: Private return deepSource error

File: packages/translation/src/lang/en.ts
Comment (hillaliy) [reply packages/translation/src/lang/en.ts:1103]: I use colons. If I add the colons in the code, then in left-to-right languages ​​they will remain from left to right.

File: packages/translation/src/lang/en.ts
Comment (hillaliy) [reply packages/translation/src/lang/en.ts:1097]: minute: "1 minute:",
        minutes: "{count} minutes:",

For 1 I should use the singular minute

File: packages/widgets/src/health-monitoring/component.tsx
Comment (hillaliy) [reply packages/widgets/src/health-monitoring/component.tsx:143]: Because I use singular for minute I don't think I need to use map for 2 items (5, 15)

**Comment by SeDemal** [comment]
> > Screen.Recording.2024-09-20.at.12.34.51.mov
> > Thanks guys for all comments. I have learned a lot from this Integration / Widget.
> 
> That's a lot better, good improvement, congrats on getting there, and thanks for sticking with it. I'll need to take a very close look at the code though seeing what Mani said.
> 
> Regarding the class-naming, Although it's tedious and ugly in the code, when I said to name every element, I meant it. This is really important if we want to make CSS as easy for users as possible, especially those that know nothing about it. We want to be able to give a name and directly affect it with CSS. Straight to the point.
> 
> For advanced information, we're now doing them in modals as hover cards don't scale nicely. Those don't need class-naming though so you're fine there.
> 
> If it's as simple as a title though, just do a Tooltip.
> 
> in recap, if it's not a Tooltip or inside a modal, put a className on it.

This comment is still relevant to what needs to be fixed in my opinion.
Please see your discord messages if you need help on this.

**Review by manuel-rw — DISMISSED** [review]
Looks good to me. @hillaliy please wait for at least another approval before merging. @SeDemal @Meierschlumpf please have another look

File: packages/api/src/router/widgets/health-monitoring.ts
Comment (Meierschlumpf) [reply packages/api/src/router/widgets/health-monitoring.ts:19]: No, it probably better to deal with `null` here

File: packages/api/src/router/widgets/health-monitoring.ts
Comment (Meierschlumpf) [inline packages/api/src/router/widgets/health-monitoring.ts:22]: Also, please remove fields like `decryptedSecrets` from the return statement as we don't want to send them to the frontend

**Review by Meierschlumpf — CHANGES_REQUESTED** [review]

File: packages/api/src/router/widgets/health-monitoring.ts
Comment (hillaliy) [reply packages/api/src/router/widgets/health-monitoring.ts:19]: if I return null I get a lot of type error from component.

**Review by manuel-rw — APPROVED** [review]
Please wait for a second approval

**Review by Meierschlumpf — APPROVED** [review]

File: packages/widgets/src/health-monitoring/component.tsx
Comment (SeDemal) [inline packages/widgets/src/health-monitoring/component.tsx:45]: Only missing subscription then we're good.
PM me, if you still need help on that.

**Pull request merged**
